File: DS-Algorithms/Searching/jump_search/code.c
#include<stdio.h>
#include<time.h>
#include<stdlib.h>
#include<math.h>

int min(int a, int b){
    return (a>b)?b:a;
}
int jumpSearch(int arr[], int arr_size, int element){
    int step = sqrt(arr_size);
    int prev = 0;
    while(arr[min(step,arr_size)-1] < element){
        prev = step;
        step+=sqrt(arr_size);
        if(prev >= arr_size)return -1;
    }
    while(arr[prev] < element){
        prev++;
        if(prev == min(step,arr_size))return -1;
    }
    if(arr[prev] == element)return prev;
    return -1;
}

int main(){
    //Code for randomly creating an array
    srand(time(0));
    int arr_size = (rand()%50000 + 50); //Means size lie between 50 to 149
    int i;
    int *arr = (int*)malloc(sizeof(int)*arr_size);
    int lowerLimit = 10;
    int maxLimit = 50;
    for(i = 0; i < arr_size; ++i){
        arr[i] = (rand()%maxLimit)+lowerLimit;
        lowerLimit = arr[i];
        
    }
    printf("%d\n",jumpSearch(arr,arr_size,arr[rand()%arr_size]+1));
    return 0;

}

File: DS-Algorithms/Searching/jump_search/code.py
import random
import math

def min(a,b):
    if(a>b):
        return b
    else:
        return a

def jump_search(arr, element):
    arr_len = len(arr)
    step = int(math.floor(math.sqrt(arr_len)))
    prev = 0
    while arr[min(step,arr_len)] < element:
        prev = step
        step = step + int(math.floor(math.sqrt(arr_len)))
        if prev >= arr_len:
            return -1
    while arr[prev]<element:
        prev = prev + 1
        if prev == min(step,arr_len):
            return -1
    if arr[prev] == element:
        return prev
    return -1
arr_length = random.randint(50,5000)
arr = []
i = 0
lowerLimit = 10
maxLimit = 50
while i < arr_length:
    arr.append(random.randint(lowerLimit,maxLimit))
    lowerLimit = arr[i]
    maxLimit = lowerLimit + 50
    i = i + 1
print(jump_search(arr,arr[random.randint(0,arr_length)]))